TDataFrame.hxx

```diff
@@ -22,828 +22,1042 @@
-   virtual void CreateSlots(unsigned nSlots) = 0;
-   virtual std::string GetName() const = 0;
-   virtual void* GetValue(unsigned slot, int entry) = 0;
-   virtual const std::type_info& GetTypeId() const = 0;
-};
-using TmpBranchBasePtr = std::unique_ptr<TDataFrameBranchBase>;
-using TmpBranchBaseVec = std::vector<TmpBranchBasePtr>;
-
-
-// Forward declaration
-class TDataFrame;
-
-// Smart pointer for the return type of actions
-class TActionResultPtrBase {
-   friend TDataFrame;
-   TDataFrame& fFirstData;
-   std::shared_ptr<bool> fReadyPtr = std::make_shared<bool>(false);
-   void SetReady() {
-      *fReadyPtr = true;
-   }
-protected:
-   bool IsReady() {
-      return *fReadyPtr;
-   }
-   void TriggerRun();
-public:
-   TActionResultPtrBase(TDataFrame& firstData);
-};
-
-using TActionResultPtrBaseVec = std::vector<TActionResultPtrBase*>;
-
-template<typename T>
-class TActionResultPtr : public TActionResultPtrBase{
-   std::shared_ptr<T> fObjPtr;
-public:
-   TActionResultPtr(std::shared_ptr<T> objPtr, TDataFrame& firstData)
-      : TActionResultPtrBase(firstData), fObjPtr(objPtr) {}
-   T *get() {
-      if (!IsReady()) TriggerRun();
-      return fObjPtr.get();
-   }
-   T& operator*() { return *get(); }
-   T *operator->() { return get(); }
-   T *getUnchecked() { return fObjPtr.get(); }
+   virtual ~TDataFrameActionBase() {}
+   virtual void Run(unsigned int slot, int entry) = 0;
+   virtual void BuildReaderValues(TTreeReader &r, unsigned int slot) = 0;
+   virtual void CreateSlots(unsigned int nSlots) = 0;
 };
 
+using ActionBasePtr_t = std::shared_ptr<TDataFrameActionBase>;
+using ActionBaseVec_t = std::vector<ActionBasePtr_t>;
 
 // Forward declarations
-template<int S, typename T>
-T& GetBranchValue(TVBPtr& readerValue, unsigned slot, int entry, const std::string& branch, TDataFrame& df);
+template <int S, typename T>
+T &GetBranchValue(TVBPtr_t &readerValues, unsigned int slot, int entry, const std::string &branch,
+                  std::weak_ptr<Details::TDataFrameImpl> df);
 
-
-template<typename F, typename PrevDataFrame>
+template <typename F, typename PrevDataFrame>
 class TDataFrameAction : public TDataFrameActionBase {
-   using BranchTypes = typename RemoveFirst<typename FunctionTraits<F>::ArgTypes>::Types;
-   using TypeInd = typename gens<BranchTypes::size>::type;
+   using BranchTypes_t = typename TDFTraitsUtils::TRemoveFirst<typename TDFTraitsUtils::TFunctionTraits<F>::ArgTypes_t>::Types_t;
+   using TypeInd_t = typename TDFTraitsUtils::TGenS<BranchTypes_t::fgSize>::Type_t;
 
 public:
-   TDataFrameAction(F f, const BranchList& bl, PrevDataFrame& pd)
-      : fAction(f), fBranches(bl), fTmpBranches(pd.fTmpBranches),
-        fPrevData(pd), fFirstData(pd.fFirstData) { }
+   TDataFrameAction(F f, const BranchList &bl, std::weak_ptr<PrevDataFrame> pd)
+      : fAction(f), fBranches(bl), fTmpBranches(pd.lock()->fTmpBranches), fPrevData(pd.lock().get()),
```

Comment:
potentially dangerous to lock the weak pointer without checking its validity. I would suggest to take a `shared_ptr` in the constructor, so at construction time we are sure the previous object in the functional chain exists, and then store it in a `weak_ptr`. same for `TDataFrameFilter` and `TDataFrameBranch`